### ISSUE TITLE
Register Default TypeConverters through .net mechanism

### ISSUE THREAD

**Issue by tisis2** [AvaloniaUI/Avalonia#12194]
**Is your feature request related to a problem? Please describe.**
as far as i see there is no way to find the default converters for IImage or FontFamily and others during runtime. in WPF all type convertions are registered through the TypeConverterAttribute. that leads to the fact that you can find the converter also during runtime. e.g. in the past i created a converter for WPF that would format a string with bound values and should be used as it would have been a input from xaml. since the type converters are not used when using bindings i created a converter that would do string format and convert to target type. for reference here:
```
public class StringFormatToTargetTypeConverter : IValueConverter, IMultiValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return Convert(new object[] { value }, targetType, parameter, culture);
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        throw new NotImplementedException();
    }

    [Localizable(false)]
    public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
    {
        if (parameter is null)
        {
            throw new ArgumentException("string format set as parameter");
        }
        
        var typeConverter = TypeDescriptor.GetConverter(targetType);
        if (typeConverter is null)
        {
            throw new ArgumentException($"no TypeConverter for target type found: {targetType.FullName}");
        }

        if (!typeConverter.CanConvertFrom(typeof(string)))
        {
            throw new ArgumentException($"TypeConverter '{typeConverter.GetType().FullName}' cannot convert from string");
        }

        var formatted = string.Format(parameter.ToString()!, values);
        return typeConverter.ConvertFrom(formatted);
    }

    public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
    {
        throw new NotImplementedException();
    }
}
```
this is not possible in Avalonia since `TypeDescriptor.GetConverter(targetType)` would not always find the converter (e.g. BitmapTypeConverter for IImage). instead the type convertion for these type are done in the xaml loader here https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguage.cs#L128. this makes it hard to write very generic converters like the above example. it would be needed to copy this mapping over to your project and resolve the correct converter on your own. what could later on lead to incompatibilities on future update when the list or logic changes.

**Describe the solution you'd like**
use the typeconvertion mechanism of .net or provide a way to fetch the converter used for a type

**Comment by maxkatz6** [comment]
Correct, in Avalonia we either skip type converters at all, injecting type parsing/casting via a compiler or something like hardcoding these converters in the compiler like BitmapTypeConverter.

In general problem with TypeDescriptor is that it often relies on reflection.

I think we have two options:
1. Add type converter on all of these missing members (there is more of them, see what we handle compile time without converters: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguageParseIntrinsics.cs)
2. Or make sure that `DefaultValueConverter` works as a replacement. It should already work for most types, but not sure about bitmaps (these could be a special case as well in your converter).

**Comment by fitdev** [comment]
@maxkatz6 

Sorry to hijack the issue, but just to clarify...

If I have a custom type `MyCustomColorType`, and I want its value to be bindable in XAML in places where say Avalonia's `IBrush` is expected like `<Border Background="{Binding ColorValueTypesAsMyCustomColorType}" />`, currently there is no way to do so unless a `Converter` is also specified with _every binding in XAML markup_?

I.e. there is no way to somehow "register" with Avalonia the notion that `MyCustomColorType` is convertible to `IBrush` via a given function / `IValueConverter` / `TypeConverter` / some other mechanism?

Is this a limitation of _compiled_ bindings or all bindings (including reflection bindings)?

If currently it is indeed not possible, then perhaps some helper facility can be introduced in Avalonia:

- It can work with both compiled and reflection bindings
- At XAML compile time can be used in much the same way as `IValueConverters` provided in the markup via `Converter` parameter are currently used.
- Of course users will be required at runtime to manually call type registration functions like `ValueConverterRegistry.RegisterConverter<TFrom, TTo>( ... )` for the binding to actually work (or else it should default to default behavior, i.e. as if no `Converter` was specified)

**Comment by VladimirDrobyshev** [comment]
@maxkatz6

In my opinion, we need the ability to specify a TypeConverter attribute for the property.

For example, it is not possible to define a custom property in xaml that is of type System.Numeric.Vector3 or other similar struct without a converter.
Avalonia has default converters for the Color and Thickness, but custom controls sometimes require properties of custom or system structs types.

**Comment by maxkatz6** [comment]
@VladimirDrobyshev isn't it already possible?
I.e. with code:
```
[TypeConverter(typeof(MyVectorConverter))]
public Vector3 Value { get; set; }
```
XAML should use this converter instead. If not, it's a reasonable feature proposal then.

**Comment by fitdev** [comment]
@maxkatz6 
Even if it is possible (not sure if it works with Compile Bindings), it is still a bit repetitive/verbose. It would be great to have the ability to just make Avalonia aware of default conversions to apply like I suggested in the previous comment. This avoids needless repetition for a set of types that should always be converted the same way.

**Comment by kekekeks** [comment]
I think for to work properly with compile-time resolution we'd need our own attribute that defines both converted and supported target type.
e. g. 
```cs
[XamlTypeConverter(typeof(IImage), typeof(MyClassImageConverter)]
class MyClass{}
```

If we blindly check for `[TypeConverter]` on the source type, we'd be assuming that said source type is convertible into _anything_ which will lead to run-time errors later.

**Comment by VladimirDrobyshev** [comment]
Sorry for the long pause.

No, it doesn't work now, i get compile time error:

> Error AVLN3000 Avalonia: Unable to find suitable setter or adder for property VectorProperty of type AvaloniaTest.Material for argument System.Runtime:System.String, available setter parameter lists are:
> Avalonia.UnsetValueType
> Avalonia.Data.IBinding
> System.Numerics.Vector3 Line 15, position 59.

**Comment by beto-rodriguez** [comment]
```c#
[TypeConverter(typeof(MyVectorConverter))]
public Vector3 Value { get; set; }
```

I second this feature, it seems impossible to specify a custom converter for a property.

**Comment by beto-rodriguez** [comment]
Additional information:

It seems that the TypeConverter attribute is partially supported/bugged, in my case for the following properties, it works only in one property.

```c#
public static readonly new global::Avalonia.AvaloniaProperty<LiveChartsCore.Painting.Paint?> CrosshairLabelsPaintProperty;
[System.ComponentModel.TypeConverter(typeof(HexToPaintTypeConverter))]
public new LiveChartsCore.Painting.Paint? CrosshairLabelsPaint
{
    get => _baseType.CrosshairLabelsPaint;
    set => SetValue(CrosshairLabelsPaintProperty, value);
}

public static readonly new global::Avalonia.AvaloniaProperty<LiveChartsCore.Drawing.LvcColor?> CrosshairLabelsBackgroundProperty;
[System.ComponentModel.TypeConverter(typeof(HexToLvcColorTypeConverter))]
public new LiveChartsCore.Drawing.LvcColor? CrosshairLabelsBackground
{
    get => _baseType.CrosshairLabelsBackground;
    set => SetValue(CrosshairLabelsBackgroundProperty, value);
}
```

The `CrosshairLabelsBackground` property works as expected, the type converter is parsed properly to `LvcColor`, this is a value type, that is maybe why it works?

But for the CrosshairLabelsPaint, which is a reference type, i get the next error:

![Image](https://github.com/user-attachments/assets/7431a485-93d9-4da5-a6a5-1e028221b338)

As a workaround, I build a XAML extension to parse the string, but it would be nice to fix this feature.

**Comment by beto-rodriguez** [comment]
**Update again**

Implementing a static Parse method on the type of the property will just work, in my case my LvcColor had this method, and my Paint class didn't. I added the method and now the XAML is parsed to the target type.

I really can't find docs about this on WPF/WinUI or Avalonia, not even copilot was able to... but it works.

AI chat:

![Image](https://github.com/user-attachments/assets/f93ef0ae-5780-411f-b107-25e8587e8fb8)

![Image](https://github.com/user-attachments/assets/6dfcdee2-4d94-4fab-bb02-3ad3cf08ddf7)

**Comment by racerxdl** [comment]
Why not allow users to change the DefaultValueConverter.Instance field? (instead making it readonly).

Then at least users can extend it.

**Comment by maxkatz6** [comment]
There is no guarantee that value converter will get used. I can repeat that Avalonia attempts to parse all built-in types in compile time. `TypeConverter` is the opposite of that, adding reflection dependency. There are discussions to extend compile-time parsing capabilities, but this issue is not about that.

Instead, this issue should be about allowing `TypeConverter` on specific properties like @beto-rodriguez and @VladimirDrobyshev mentioned.
I.e. code below should parse vector in runtime using provided converter:
```
[TypeConverter(typeof(MyVectorConverter))]
public Vector3 Value { get; set; }
```

Meanwhile `TypeConverter` on a custom type is already supported:
```
[TypeConverter(typeof(MyVectorConverter))]
public class Vector
{
}
```

**Comment by racerxdl** [comment]
Just as a side note which might be important context: I recently had to do a hack to override the `DefaultValueConverter.Instance` to be able to change the default method converter (we have a custom Task type, and Avalonia/Reactive uses C# Task class instead of its interfaces) because of the `MethodToCommandConverter`.

So it might be good to consider being able to override any default converters as well.
